### ISSUE TITLE
lucene-expressions fail at compiling a simple script

### ISSUE THREAD

**Issue by jpountz** [elastic/elasticsearch#13949]
If you look at this build failure: http://build-us-00.elastic.co/job/es_core_master_regression/3629/testReport/junit/org.elasticsearch.script.expression/MoreExpressionTests/testSpecialValueVariable/ Elasticsearch failed to compile an aggregation script on one shard (only one) due to the following error:

```
ScriptException[Failed to compile inline script [_value * 3] using lang [expression]]; nested: ArrayIndexOutOfBoundsException[7];
	at org.elasticsearch.script.ScriptService.compileInternal(ScriptService.java:312)
	at org.elasticsearch.script.ScriptService.compile(ScriptService.java:256)
	at org.elasticsearch.script.ScriptService.search(ScriptService.java:455)
	at org.elasticsearch.search.aggregations.support.ValuesSourceParser.createScript(ValuesSourceParser.java:230)
	at org.elasticsearch.search.aggregations.support.ValuesSourceParser.config(ValuesSourceParser.java:224)
	at org.elasticsearch.search.aggregations.metrics.NumericValuesSourceMetricsAggregatorParser.parse(NumericValuesSourceMetricsAggregatorParser.java:66)
	at org.elasticsearch.search.aggregations.AggregatorParsers.parseAggregators(AggregatorParsers.java:198)
	at org.elasticsearch.search.aggregations.AggregatorParsers.parseAggregators(AggregatorParsers.java:103)
	at org.elasticsearch.search.aggregations.AggregationParseElement.parse(AggregationParseElement.java:60)
	at org.elasticsearch.search.SearchService.parseSource(SearchService.java:752)
	... 10 more
Caused by: java.lang.ArrayIndexOutOfBoundsException: 7
	at org.antlr.v4.runtime.atn.ParserATNSimulator.addDFAEdge(ParserATNSimulator.java:1956)
	at org.antlr.v4.runtime.atn.ParserATNSimulator.computeTargetState(ParserATNSimulator.java:609)
	at org.antlr.v4.runtime.atn.ParserATNSimulator.execATN(ParserATNSimulator.java:485)
	at org.antlr.v4.runtime.atn.ParserATNSimulator.adaptivePredict(ParserATNSimulator.java:424)
	at org.apache.lucene.expressions.js.JavascriptParser.expression(JavascriptParser.java:494)
	at org.apache.lucene.expressions.js.JavascriptParser.compile(JavascriptParser.java:113)
	at org.apache.lucene.expressions.js.JavascriptCompiler.getAntlrParseTree(JavascriptCompiler.java:216)
	at org.apache.lucene.expressions.js.JavascriptCompiler.compileExpression(JavascriptCompiler.java:189)
	at org.apache.lucene.expressions.js.JavascriptCompiler.compile(JavascriptCompiler.java:122)
	at org.elasticsearch.script.expression.ExpressionScriptEngineService$1.run(ExpressionScriptEngineService.java:107)
	at org.elasticsearch.script.expression.ExpressionScriptEngineService$1.run(ExpressionScriptEngineService.java:102)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.elasticsearch.script.expression.ExpressionScriptEngineService.compile(ExpressionScriptEngineService.java:102)
	at org.elasticsearch.script.ScriptService.compileInternal(ScriptService.java:310)
```

The failure does not reproduce so I'm wondering there might be a concurrency issue here?

**Comment by rmuir** [comment]
I think its this bug: https://github.com/antlr/antlr4/issues/804

Lets upgrade to antlr 4.5.1, please fix this in lucene!

**Comment by clintongormley** [comment]
@jdconrad see ^^

**Comment by jdconrad** [comment]
Just saw this.  Will take a look at fixing this tonight.

**Comment by jdconrad** [comment]
@rmuir, Good find on the bug.  Thanks!  Attached a patch here (https://issues.apache.org/jira/browse/LUCENE-6830) for Lucene.

**Comment by clintongormley** [comment]
Fixed in Lucene 5.4
